### PR TITLE
Polls/Voting Feature in lib-jitsi-meet

### PR DESCRIPTION
**Pull request by mmoanis**
Related to [#3827](https://github.com/jitsi/jitsi-meet/pull/3827)

Implement API in lib-jitsi-meet for managing a poll inside the conference room. API is not documented for now.

The API export 3 events to conference that represent:
1. Poll started
2. Someone voted in a poll
3. Poll ended

Additionally, the poll module keeps track of the current poll by saving its state to a prosody module and listens to messages from the module if for example a user just joined a conference room to get notified about the poll.

### CODE REVIEW

**Comment by mmoanis** [comment]
CLA signed for 8x8

File: modules/polls/polls.js [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+const MESSAGE_POLL_STARTED = 'poll-started';
+const MESSAGE_POLL_VOTE = 'poll-vote';
+
+
+/**
+ * A module for managing poll session related functionality.
+ * This module handles communication with all participants regarding the poll
+ * session management. It also keep the record of what poll is going
+ * on using the server side prosody-module if it is present.
+ */
```
Comment (virtuacoplenny) [inline modules/polls/polls.js]: I really appreciate how you tried to encapsulate polls.

File: modules/polls/polls.js [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+     * @param {string} roomJid - Room JID
+     * @param {Object} poll - Object containing the information about the poll.
+     * @param {Object} choices - Object containing poll choices by their ID.
+     * @param {Object} question - Object with information about
+     * the question asked.
+     * @returns {void}
+     */
+    startPoll(roomjid, poll, choices, question) {
+        if (!poll || !choices || !question) {
+            return;
```
Comment (virtuacoplenny) [inline modules/polls/polls.js]: Should an error be logged here?

File: modules/polls/polls.js [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        }
+
+        if (this.poll && this.choices && this.question) {
+            logger.warn('Attempt to create a poll '
+                + 'while one is currently active. Please call endPoll first');
+
+            return;
+        }
+
+        logger.log(`Poll Start: ${poll} ${choices} ${question}`);
```
Comment (virtuacoplenny) [inline modules/polls/polls.js]: I'm uncertain how I feel about logging potentially private details.

File: modules/polls/polls.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            question
+        );
+    }
+
+    /**
+     * Emit user voted event.
+     * @param {string} choiceID - Choice ID.
+     * @param {string} userID - ID of user who voted.
+     */
+    _doVote(choiceID, userID) {
```
Comment (virtuacoplenny) [inline modules/polls/polls.js]: Maybe this should be a toggleVote? I don't feel strongly about it though.

Comment (mmoanis) [reply modules/polls/polls.js]: Yeah, its better

File: modules/polls/polls.js (hunk 3 above)
Comment (mmoanis) [reply modules/polls/polls.js]: fixed

File: modules/polls/polls.js (hunk 1 above)
Comment (mmoanis) [reply modules/polls/polls.js]: speaking of such, currently assigning IDs for example happens actually in the UI. I think it should be moved down here, right? Is there a way to generate unique IDs in lib-jitsi-meet. I didn't want to add a new dependency package.

**Comment by mmoanis** [comment]
Fixed review about logging and function names.

I would like to ask about the possibility of generating unique IDs inside lib-jitsi-meet because I think building the data structure that represent the poll should be here not in jitsi-meet.

File: modules/polls/polls.js (hunk 1 above)
Comment (virtuacoplenny) [reply modules/polls/polls.js]: What about prosody? Could that store the source of truth for all polls in the muc? It has uuid capabilities. Boris will also be reviewing your PRs and he's more knowledge about that side of the world.

**Comment by virtuacoplenny** [comment]
Jenkins test this please.

File: modules/polls/polls.js (hunk 1 above)
Comment (mmoanis) [reply modules/polls/polls.js]: Yeah currently prosody module store that but it only stores it and tell people when they join about it.

**Comment by kneza23** [comment]
Any progress on this?

**Comment by please-wait** [comment]
I believe the change is tied to this in Jitsi-meet: https://github.com/jitsi/jitsi-meet/pull/3827

It would be a very cool feature to have.

**Comment by Aaryan-kapur** [comment]
Was this integrated @mmoanis

**Comment by saghul** [comment]
https://github.com/jitsi/jitsi-meet/pull/3827#issuecomment-698333629

**Pull request closed without merging**